search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt

```diff
@@ -0,0 +1,144 @@
+import com.egm.stellio.shared.util.JsonLdUtils.JSONLD_ID_TERM
+import com.egm.stellio.shared.util.JsonLdUtils.JSONLD_TYPE_TERM
+import com.egm.stellio.shared.util.JsonLdUtils.NGSILD_CREATED_AT_TERM
+import com.egm.stellio.shared.util.JsonLdUtils.NGSILD_DATASET_ID_TERM
+import com.egm.stellio.shared.util.JsonLdUtils.NGSILD_MODIFIED_AT_TERM
+import com.egm.stellio.shared.util.JsonLdUtils.NGSILD_OBSERVED_AT_TERM
+import com.egm.stellio.shared.util.JsonLdUtils.NGSILD_SCOPE_TERM
+import java.time.ZonedDateTime
+import kotlin.random.Random.Default.nextBoolean
+
+typealias CompactedEntityWithCSR = Pair<CompactedEntity, ContextSourceRegistration>
+typealias DataSetId = String?
+typealias AttributeByDataSetId = Map<DataSetId, CompactedAttributeInstance>
+object ContextSourceUtils {
+
+    fun mergeEntities(
+        localEntity: CompactedEntity?,
+        remoteEntitiesWithMode: List<CompactedEntityWithCSR>
+    ): IorNel<NGSILDWarning, CompactedEntity?> = iorNel {
+        if (localEntity == null && remoteEntitiesWithMode.isEmpty()) return@iorNel null
+
+        val mergedEntity: MutableMap<String, Any> = localEntity?.toMutableMap() ?: mutableMapOf()
+
+        remoteEntitiesWithMode.sortedBy { (_, csr) -> csr.isAuxiliary() }
+            .forEach { (entity, csr) ->
+                mergedEntity.putAll(
+                    getMergeNewValues(mergedEntity, entity, csr).toIor().toIorNel().bind()
+                )
+            }
+
+        return@iorNel mergedEntity.toMap()
+    }
+
+    private fun getMergeNewValues(
+        localEntity: CompactedEntity,
+        remoteEntity: CompactedEntity,
+        csr: ContextSourceRegistration
+    ): Either<NGSILDWarning, CompactedEntity> = either {
+        remoteEntity.mapValues { (key, value) ->
+            val localValue = localEntity[key]
+            when {
+                localValue == null -> value
+                key == JSONLD_ID_TERM || key == JSONLD_CONTEXT -> localValue
+                key == JSONLD_TYPE_TERM || key == NGSILD_SCOPE_TERM ->
+                    mergeTypeOrScope(localValue, value)
+                key == NGSILD_CREATED_AT_TERM ->
+                    if ((value as String?).isBefore(localValue as String?)) value
+                    else localValue
+                key == NGSILD_MODIFIED_AT_TERM ->
+                    if ((localValue as String?).isBefore(value as String?)) value
+                    else localValue
+                else -> mergeAttribute(
+                    localValue,
+                    value,
+                    csr
+                ).bind()
+            }
+        }
+    }
+
+    fun mergeTypeOrScope(
+        currentValue: Any, // String || List<String> || Set<String>
+        remoteValue: Any
+    ) = when {
+        currentValue == remoteValue -> currentValue
+        currentValue is List<*> && remoteValue is List<*> -> (currentValue.toSet() + remoteValue.toSet()).toList()
+        currentValue is List<*> -> (currentValue.toSet() + remoteValue).toList()
```

Comment:
since both `List` and `Set` inherit from `Iterable`, you could work with this type and avoid the `List` <-> `Set` transformations and make this a bit more digest ;)

Reply:
I could replace it by :
```
 currentValue == remoteValue -> currentValue
        currentValue is Iterable<*> && remoteValue is List<*> -> (currentValue.toSet() + remoteValue.toSet())
        currentValue is Iterable<*> -> (currentValue.toSet() + remoteValue)
        remoteValue is Iterable<*> -> (remoteValue.toSet() + currentValue)
        else -> listOf(currentValue, remoteValue)
```
But at the end it is no longer a list and it could impact the rest of the code.

Reply:
as you know, not really fan of this function but not a big deal. let's focus on more important things.